python/dolfinx/fem/function.py

```diff
@@ -120,80 +121,86 @@ def __init__(self, ufl_expression: ufl.core.expr.Expr, x: np.ndarray,
         # Setup data (evaluation points, coefficients, constants, mesh, value_size).
         # Tabulation function.
         ffi = cffi.FFI()
-        fn = ffi.cast("uintptr_t", self.ufcx_expression.tabulate_expression)
-
-        value_size = ufl.product(self.ufl_expression.ufl_shape)
 
-        ufl_coefficients = ufl.algorithms.extract_coefficients(ufl_expression)
-        coefficients = [ufl_coefficient._cpp_object for ufl_coefficient in ufl_coefficients]
+        # Prepare coefficients data. For every coefficient in form take
+        # its C++ object.
+        original_coefficients = ufl.algorithms.extract_coefficients(ufl_expression)
+        coeffs = [original_coefficients[self._ufcx_expression.original_coefficient_positions[i]]._cpp_object
+                  for i in range(self._ufcx_expression.num_coefficients)]
 
         ufl_constants = ufl.algorithms.analysis.extract_constants(ufl_expression)
-        constants = [ufl_constant._cpp_object for ufl_constant in ufl_constants]
+        constants = [constant._cpp_object for i, constant in enumerate(ufl_constants)]
 
-        # Getcpp Expression type
-        def expressiontype(dtype):
+        def create_expression(dtype):
             if dtype is np.float64:
-                return _cpp.fem.Expression_float64
+                return _cpp.fem.create_expression_float64
```

Comment:
Could probably copy DirichletBCMetaClass and FormMetaClass approach for Expression.